tensorflow/compiler/tf2tensorrt/convert/ops/resize.cc

```diff
@@ -0,0 +1,658 @@
+            "corners=False, dynamic spatial shape) is unsupported.");
+      }
+    }
+    return Status::OK();
+  }
+
+  static bool UseConvResizeWorkaround(const ResizeAttributes& attrs) {
+    if (!attrs.align_corners && attrs.mode == nvinfer1::ResizeMode::kLINEAR) {
+      if (!IS_TRT_VERSION_GE(8, 0, 0, 0)) {
+        return true;
+      }
+
+      int64_t force_bilinear_resize_war = 0;
+      ReadInt64FromEnvVar("TF_TRT_FORCE_BILINEAR_RESIZE_WAR", 0,
+                          &force_bilinear_resize_war);
+      if (force_bilinear_resize_war > 0) {
+        LOG(WARNING) << "Forcing the use of LHS-dilated convolution WAR for "
+                        "ResizeBilinear when aligned_corners=false";
+        return true;
+      }
+    }
+    return false;
+  }
+
+  Status Convert() {
+    // Transpose tensor from NHWC to NCHW format. Bilinear resize assumes you
+    // are in this mode.
+    TF_RETURN_IF_ERROR(params_->converter->TransposeTensor(
+        input_tensor_, {0, 3, 1, 2}, &input_tensor_, params_->node_def,
+        "to_NCHW"));
+
+    StatusOr<TRTNetworkBuilder> builder = TRTNetworkBuilder::Create(
+        params_->converter->network(), params_->weight_store);
+    TRT_ENSURE_OK(builder);
+
+    // Calculate the output shape as static dimensions or a shape tensor:
+    // Given input shape [N, C, H, W] and output size [H_out, W_out],
+    // output shape equals [N, C, H_out, W_out].
+    TRT_TensorOrWeights size_input = params_->inputs.at(1);
+
+    ITensorProxyPtr output;
+    // For TensorRT < 8, we require a workaround for (bilinear,
+    // align_corners=false). Otherwise, create a single TRT IResizeLayer.
+    if (UseConvResizeWorkaround(attrs_)) {
+      StatusOr<nvinfer1::ITensor*> out =
+          ResizeUsingConvAndDilation(input_tensor_->trt_tensor());
+      TRT_ENSURE_PTR_OK(out);
+      output = *out;
+    } else {
+      // Add resize layer.
+      nvinfer1::IResizeLayer* layer = params_->converter->network()->addResize(
+          *input_tensor_->trt_tensor());
+      TRT_ENSURE(layer);
+      params_->converter->SetLayerName(layer, params_->node_def);
+
+      // Set layer parameters.
+      layer->setResizeMode(attrs_.mode);
+      layer->setAlignCorners(attrs_.align_corners);
+      TF_RETURN_IF_ERROR(SetCoordinateTransform(layer));
+      TF_RETURN_IF_ERROR(ComputeAndSetOutputShape(
+          input_tensor_, params_->inputs.at(1), &*builder, layer,
+          attrs_.has_static_input_shape, attrs_.has_static_output_shape));
+      output = layer->getOutput(0);
+    }
+
+    // Get output tensor. Transpose it from NCHW to NHWC.
+    TF_RETURN_IF_ERROR(params_->converter->TransposeTensor(
+        output, {0, 2, 3, 1}, &output, params_->node_def, "to_NHWC"));
+    this->AddOutput(output);
+    return Status::OK();
+  }
+
+  // For TensorRT >= 8, all possible attribute combinations are directly
+  // supported.
+  Status SetCoordinateTransform(nvinfer1::IResizeLayer* layer) const {
+#if IS_TRT_VERSION_GE(8, 0, 0, 0)
```

Comment:
Will this contradict with "testing the workaround in TensorRT 8"?

Reply:
No, the functions to set the coordinate transform are for the direct TRT path in TensorRT 8. That API is new in TensorRT8.

Reply:
SetCoordinateTransform is not called in the direct conversion path. You can see I added an environment variable `TF_TRT_FORCE_BILINEAR_RESIZE_WAR`. If that variable is set to a non-zero value, then even in TensorRT 8, the converter will take the workaround path when `align_corners=false`, even though it is directly supported by  the `IResizeLayer` in TRT >= 8.0.